topics/introduction/tutorials/python-advanced-np-pd/tutorial.md

```diff
@@ -0,0 +1,988 @@
+FBgn0038261   1.525142e-04      CG14856
+FBgn0039178   1.629061e-04       CG6356
+FBgn0034636   1.951192e-04      CG10440
+
+[130 rows x 2 columns]
+```
+```python
+mask = subset.loc[:, 'P-adj'] < 0.000005
+new_data = subset[mask]
+print(new_data)
+```
+```output
+
+                     P-adj    Gene name
+GeneID                                 
+FBgn0039155  1.387207e-203         Kal1
+FBgn0003360  4.039734e-178         sesB
+FBgn0026562  8.149380e-172  BM-40-SPARC
+FBgn0025111  6.105789e-164         Ant2
+FBgn0029167  2.698811e-111          Hml
+...                    ...          ...
+FBgn0039593   1.544012e-07       CG9989
+FBgn0265512   3.413938e-07          mlt
+FBgn0030326   3.625857e-07       CG2444
+FBgn0039485   7.174821e-07      CG17189
+FBgn0025836   1.071884e-06     RhoGAP1A
+
+[113 rows x 2 columns]
+```
+
+If we have not had specified the column, that the expression should be applied to, then it would have been applied to the entire dataframe. But the dataframe contains different type of data. In that case, an error would occur. 
+
+Consider the following example of a dataframe consisting only of numerical data. The expression and the mask would be normally applied to the data and the mask would return `NaN` for the data that don't comply with the expression. 
+
+```python
+subset = data.loc[:, ['StdErr',	'Wald-Stats', 'P-value', 'P-adj']]
+mask = subset < 0.05
+new_data = subset[mask]
+print(new_data)
+```
+```output
+             StdErr  Wald-Stats        P-value          P-adj
+GeneID                                                       
+FBgn0039155     NaN  -30.740913  1.617357e-207  1.387207e-203
+FBgn0003360     NaN  -28.748637  9.419922e-182  4.039734e-178
+FBgn0026562     NaN  -28.225437  2.850430e-175  8.149380e-172
+FBgn0025111     NaN         NaN  2.847517e-167  6.105789e-164
+FBgn0029167     NaN  -22.745964  1.573284e-114  2.698811e-111
+...             ...         ...            ...            ...
+FBgn0035710     NaN         NaN   6.682473e-06   1.436480e-04
+FBgn0035523     NaN         NaN   7.103599e-06   1.523189e-04
+FBgn0038261     NaN         NaN   7.130487e-06   1.525142e-04
+FBgn0039178     NaN         NaN   7.654328e-06   1.629061e-04
+FBgn0034636     NaN   -4.430118   9.418135e-06   1.951192e-04
+
+[130 rows x 4 columns]
+```
+This is very useful because NaNs are ignored by operations like max, min, average, etc.
+
+```python
+print(new_data.describe())
+```
+```output
+       StdErr  Wald-Stats        P-value          P-adj
+count     0.0   70.000000   1.300000e+02   1.300000e+02
+mean      NaN  -11.035514   4.200552e-07   9.321175e-06
+std       NaN    5.695077   1.531253e-06   3.278527e-05
+min       NaN  -30.740913  1.617357e-207  1.387207e-203
+25%       NaN  -13.059226   4.952408e-31   1.275950e-28
+50%       NaN   -9.971703   2.109219e-19   2.563262e-17
+75%       NaN   -6.963977   6.731953e-12   3.706020e-10
+max       NaN   -4.430118   9.418135e-06   1.951192e-04
+```
+
+### Exercises
```

Comment:
Could you:

- use a question/solution box here
- not set `python` on the code blocks as it confuses the notebook conversion library

E.g https://github.com/galaxyproject/training-material/blob/main/topics/data-science/tutorials/sql-basic/tutorial.md?plain=1#L310

Reply:
Should I use a question/solution box only in this example of code? Or in all of them? Because you told me to remove the output blocks and I am a bit confused

Reply:
Ahh sorry for the confusion!

- No output for normal code blocks (the ones that are left aligned in the file / have no leading space) as those will become executable cells
- But within Question / Solution boxes, keep the outputs because they're necessary as the solutions.

Question/Solution should just be used wherever you have Exercise + Solution Currently, but not for the other code blocks that students will run like the lines above this exercise

Reply:
Ok. Thank you for the clarification